### PR TITLE
Added metabolite annotation tests and a test set for control

### PR DESCRIPTION
**Pull request by vm-vh**
Added metabolite annotation tests in the file src/memote.jl and a test set using the e.coli core json model in test/memote.jl for control.

### CODE REVIEW

**Review by exaexa — COMMENTED** [review]
Hi Vincent! Thanks for contribution, this looks moreless okay, except for the tests failing. Don't worry about the formatting test for now (I can fix that later), but please make sure the normal tests run okay.

Also, in documentation we use DocStringExtensions to generate "nice" docstring headers (you might have seen the `TYPEDSIGNATURES` thing above). I'll add 1 suggestion but pls apply that systematically.

File: src/memote.jl [hunk 1]
```diff
@@ -321,3 +321,73 @@ function is_consistent(
...
+           !haskey(metabolite_annos, keyword) && push!(missing_annos[keyword], metabolite_id)
+       end
+   end
+   return missing_annos
+end
+
+
+
+"""
+Metabolite Annotation Conformity Per Database
```
Comment (exaexa) [inline src/memote.jl:361]:
```suggestion
$(TYPEDSIGNATURES)

Metabolite Annotation Conformity Per Database
```

File: src/memote.jl [hunk 2]
```diff
@@ -321,3 +321,73 @@ function is_consistent(
...
+i.e. matching those that are defined at https://identifiers.org/.
+Returns a dictionary with a key for every database that has annotations 
+with the corresponding value being an array of all metabolites whose annotations do not match the given regex pattern.
+"""
+
+metabolites_regex = Dict("pubchem.compound" => r"^\d+$",
+    "kegg.compound" => r"^C\d+$",
+    "seed.compound" => r"^cpd\d+$",
+    "inchi_key" => r"^[A-Z]{14}\-[A-Z]{10}(\-[A-Z])?",
+    "inchi" => r"^InChI\=1S?\/[A-Za-z0-9\.]+(\+[0-9]+)?(\/[cnpqbtmsih][A-Za-z0-9\-\+\(\)\,\/\?\;\.]+)*$",
```
Comment (exaexa) [inline src/memote.jl:372]: is there please a source for this regex? I'm not super familiar with inchi but it looks like there is a bit of magic. Also, is it just a parsing check or does it check for validity of the inchi string (as in, if it does describe an actual molecule)?

**Comment by exaexa** [comment]
PS. re the failing tests -- you might need to export the functions explicitly (somewhere around here https://github.com/LCSB-BioCore/FBCModelTests.jl/blob/master/src/FBCModelTests.jl#L17 ) or use a package prefix (`FBCModelTests.some_function_name`).

File: src/memote.jl [hunk 3]
```diff
@@ -321,3 +321,73 @@ function is_consistent(
...
 end
+
+
+
+"""
+Checks for the presence of metabolite annotations.
+Returns a vector of all metabolites without any annotations.
+"""
+
+function all_unannotated_metabolites(model::StandardModel)
```
Comment (stelmo) [inline src/memote.jl:332]: Let's just make this `::MetabolitModel` and add a TODO comment to change it to `AbstractMetabolicModel` once COBREXA 2.0 releases

File: src/memote.jl (hunk 2 above)
Comment (stelmo) [reply src/memote.jl:372]: It's probably from memote.py. But please convert to a readableregex to make it more readable... See here: https://github.com/jkrumbiegel/ReadableRegex.jl

**Review by stelmo — CHANGES_REQUESTED** [review]
Nice, really good progress! Just make the changes and rerequest review.

File: src/memote.jl [hunk 4]
```diff
@@ -321,3 +321,73 @@ function is_consistent(
...
+
+
+"""
+Iterates through a model's metabolites to check if any common biochemical databases appear in the annotations field.
+Returns a dictionary with the biochemical database names being checked for as keywords 
+and the corresponding value being a vector of all metabolites missing that database name.
+"""
+
+function unannotated_metabolites(model::StandardModel,
+    annotation_keywords = ["pubchem.compound", "kegg.compound", "seed.compound", "inchi_key", "inchi",
```
[the remark this answers is not in the document]
Comment (stelmo) [reply src/memote.jl:345]: Oh sorry, it already is, my bad! No need to change anything :)

File: src/memote.jl (hunk 2 above)
Comment (stelmo) [reply src/memote.jl:372]: hmm ReadableRegex has no deps... And makes regex statements much more readable...

Comment (exaexa) [reply src/memote.jl:372]: Ah OK, maybe the sourcing comment was just too far for me to see. If there's a big overall reference to Memote.py, let's keep it there and it's good that way I'd say.

ReadableRegex is a currently very unstable dependency so let's prevent it until absolutely necessary, this regex here is pretty much okay :D

Comment (vm-vh) [reply src/memote.jl:372]: Also what is the final verdict re: readable regex?

Comment (exaexa) [reply src/memote.jl:372]: not needed, it doesn't really improve the situation here. I'll eventually make an actual parser that checks if the inchi encodes an actual molecule (instd. of just looking like inchi)

Comment (exaexa) [reply src/memote.jl:372]: (that's also the basic reason for my original question -- mostly, it was "how come this regex is valid" :D )

**Comment by exaexa** [comment]
@vm-vh  when #20 is merged pls rebase this and add yourself to the list of authors :]

**Comment by stelmo** [comment]
you will have to rebase this PR since the structure of the code dramatically changed...

**Comment by vm-vh** [comment]
Alright I've rebased and made the changes, hopefully everything is ok now, though I'm not sure that VScode actually kept the standard final newline :/

**Comment by exaexa** [comment]
> Alright I've rebased and made the changes, hopefully everything is ok now, though I'm not sure that VScode actually kept the standard final newline :/

Tests pass, great! You can actually run the formatting yourself with the scripts from the test. If you have a unix or WSL or something, you can just run this command in the directory:  https://github.com/LCSB-BioCore/FBCModelTests.jl/blob/master/.github/workflows/pr-format.yml#L27-L36 . If that's a problem lmk, I'll just apply it myself after merging

**Comment by vm-vh** [comment]
okay I attempted to run the commands you suggested (I do have unix, linux mint specifically) lets see if that worked

**Comment by vm-vh** [comment]
nope still a newline error, weird

**Comment by exaexa** [comment]
Ah ok, almost. Now you need to run `JuliaFormatter.format()` in the directory :D

Alternatively feel free to just copypaste the diff from here:
https://github.com/LCSB-BioCore/FBCModelTests.jl/actions/runs/3260031904/jobs/5353297292#step:6:9
and apply with `git apply` (run the command and copypaste the diff to the console)

**Comment by exaexa** [comment]
:thinking: 

I guess autoformatter could work too, lemme try:

**Comment by exaexa** [comment]
oh wow wtf, it worked :D

File: src/memote.jl (hunk 3 above)
Comment (exaexa) [reply src/memote.jl]: This seems fixed to me

**Review by exaexa — APPROVED** [review]
all seems good, let's get the code in and we can pick the details later

**Comment by vm-vh** [comment]
oh nice, thank you for the help!!

**Pull request merged**